sys/foraging.go

```diff
@@ -294,47 +298,61 @@ func (s *Foraging) searching(w *ecs.World) {
 	}
 	detectionProb := 1.0 - nonDetectionProb
 
-	// TODO: shuffle foragers
-	foragerQuery := s.foragerFilterSimple.Query()
-	for foragerQuery.Next() {
-		act, patch := foragerQuery.Get()
+	// implemented the shuffling of all all searching or recruited foragers here to get closer to Netlogo "ask" primitive
+	activityQuery := s.activityFilter.Query()
+	for activityQuery.Next() {
+		act := activityQuery.Get()
+		if act.Current == activity.Searching || act.Current == activity.Recruited {
+			s.foragershuffle = append(s.foragershuffle, activityQuery.Entity())
+		}
+	}
+	s.rng.Shuffle(len(s.foragershuffle), func(i, j int) { s.foragershuffle[i], s.foragershuffle[j] = s.foragershuffle[j], s.foragershuffle[i] })
 
-		if act.Current == activity.Searching {
-			if s.rng.Float64() >= detectionProb {
-				continue
+	for _, e := range s.foragershuffle {
+		act, patch := s.foragerMapper.Get(e)
+
+		if act.Current != activity.Searching { // added this to speed up the for loop
```

Comment:
Maybe it would be more efficient to have two separate shuffled slices for searching and recruited and iterate them separately? Particularly, it would reduce the number of calls to `s.foragerMapper.Get` by half, and it would reduce branching.